src/import/load-costume.js

```diff
@@ -30,76 +30,149 @@ const loadVector_ = function (costume, costumeAsset, runtime, rotationCenter, op
+            textImageElement.src = costume.textLayerAsset.encodeDataURI();
         } else {
-            imageElement.src = src;
+            loadedOne = true;
         }
-    }).then(imageElement => {
-        // createBitmapSkin does the right thing if costume.bitmapResolution or rotationCenter are undefined...
-        costume.skinId = runtime.renderer.createBitmapSkin(imageElement, costume.bitmapResolution, rotationCenter);
-        const renderSize = runtime.renderer.getSkinSize(costume.skinId);
-        costume.size = [renderSize[0] * 2, renderSize[1] * 2]; // Actual size, since all bitmaps are resolution 2
-
-        if (!rotationCenter) {
-            rotationCenter = runtime.renderer.getSkinRotationCenter(costume.skinId);
-            // Actual rotation center, since all bitmaps are resolution 2
-            costume.rotationCenterX = rotationCenter[0] * 2;
-            costume.rotationCenterY = rotationCenter[1] * 2;
-            costume.bitmapResolution = 2;
+        baseImageElement.src = costume.asset.encodeDataURI();
+    }).then(imageElements => {
+        const [baseImageElement, textImageElement] = imageElements;
+
+        let canvas = document.createElement('canvas');
+        const scale = costume.bitmapResolution === 1 ? 2 : 1;
+        canvas.width = baseImageElement.width;
+        canvas.height = baseImageElement.height;
+
+        const ctx = canvas.getContext('2d');
+        ctx.drawImage(baseImageElement, 0, 0);
+        if (textImageElement.src) {
+            ctx.drawImage(textImageElement, 0, 0);
+        }
+        if (scale !== 1) {
+            if (!runtime.v2BitmapAdapter) {
+                return Promise.reject('No V2 Bitmap adapter present.');
+            }
+            canvas = runtime.v2BitmapAdapter.resize(canvas, canvas.width * scale, canvas.height * scale);
+        }
+
+        // By scaling, we've converted it to bitmap resolution 2
+        if (rotationCenter) {
+            rotationCenter[0] = rotationCenter[0] * scale;
+            rotationCenter[1] = rotationCenter[1] * scale;
+            costume.rotationCenterX = rotationCenter[0];
+            costume.rotationCenterY = rotationCenter[1];
+        }
+        costume.bitmapResolution = 2;
+
+        return {
+            canvas: canvas,
+            rotationCenter: rotationCenter,
+            // True if the asset matches the base layer; false if it required adjustment
+            assetMatchesBase: scale === 1 && !textImageElement.src
+        };
+    })
+        .catch(e => Promise.reject(e))
+        .finally(() => {
+            // Clean up the costume object
+            delete costume.textLayerMD5;
+            delete costume.textLayerAsset;
+        });
+};
+
+const loadBitmap_ = function (costume, runtime, rotationCenter) {
+    return fetchBitmapCanvas_(costume, runtime, rotationCenter).then(fetched => new Promise(resolve => {
+        rotationCenter = fetched.rotationCenter;
+
+        const saveAssetToStorage = function (dataURI) {
```

Comment:
I would name this function something else, like `makeAsset` or `createAsset`, since it doesn't actually "save" anything to storage. (I guess this was probably calling `cache` at some point in the past?)